File: site/lib/src/firebase/DecideFire.dart

@JS()
library decidefire;


import 'dart:async';

import 'package:decideup/src/domain/DecideError.dart';
import 'package:decideup/src/domain/LoginResult.dart';
import 'package:decideup/src/domain/User.dart';
import 'package:decideup/src/services/GroupService.dart';
import 'package:decideup/src/services/TopicService.dart';
import 'package:js/js.dart';


@JS()
abstract class DecideFireJS {
  external factory DecideFireJS();
  external void checkEmail(String email, void Function(List<dynamic>, String /* error */) callback);
  external void signup(String email, String password, String displayName, void Function(DecideFireLoginResultJS) callback);
  external void login(String email, String password, void Function(DecideFireLoginResultJS) callback);
  external void currentUser(void Function(dynamic) callback);
  external void push(String path, dynamic obj, void Function(String /*uid*/) callback);
  external void save(String path, dynamic obj, void Function() callback);
  external void get(String path, void Function(dynamic result) callback);
  external List<String> getKeysForObject(dynamic jsObject);
  external void logout();
}

@JS()
abstract class DecideFireErrorJS {
  external String code();
  external String message();
}

@JS()
abstract class DecideFireLoginResultJS {
  external dynamic user();
  external DecideFireErrorJS error();
}

@JS()
@anonymous
class GroupJS {
  external String get name;
  external String get description;

  external factory GroupJS({String name, String description, List<String> users});
}

class FireLoginResult {
  dynamic user;
  DecideError error;
}

//class

class DecideFire {
  final DecideFireJS js = DecideFireJS();
  DecideFire();

  Future<List<String>> checkEmail(String email) async {
    Completer<List<String>> completer = new Completer();
    this.js.checkEmail(email, allowInterop((methods, error) {
      if(error != "") {
        completer.completeError(error);
      } else {
        List<String> methodStrings = new List();
        methods.forEach((method) => methodStrings.add(method.toString()));
        completer.complete(methodStrings);
      }
    }));

    return completer.future;
  }

  List<String> getKeysForObject(dynamic obj) {
    return js.getKeysForObject(obj);
  }

  void logout() {
    this.js.logout();
  }

  Future<String> push(String path, dynamic obj) {
    Completer<String> completer = new Completer();
    this.js.push(path, obj, allowInterop((result) => completer.complete(result)));
    return completer.future;
  }

  Future<void> save(String path, dynamic obj) {
    Completer<void> completer = new Completer();
    this.js.save(path, obj, allowInterop(() => completer.complete()));
    return completer.future;
  }

  Future<dynamic> get(String path) {
    Completer<dynamic> completer = new Completer();
    this.js.get(path, allowInterop((result) {
      completer.complete(result);
    }));

    return completer.future;
  }

  Future<dynamic> currentUser() async {
    Completer<dynamic> completer = new Completer();
    this.js.currentUser(allowInterop((user) {
      if(user != null) {
        completer.complete(user);
      } else {
        completer.complete(null);
      }
    }));
    return completer.future;
  }

  Future<FireLoginResult> signup(String email, String password, String displayName) async {
    Completer<FireLoginResult> completer = new Completer();
    this.js.signup(email, password, displayName, allowInterop((jsResult)  {
      if(jsResult.user() != null) {
        var result = FireLoginResult();
        result.user = jsResult.user();
        completer.complete(result);
      } else {
        var result = FireLoginResult();
        result.error = DecideError(jsResult.error().code(), jsResult.error().message());
        completer.complete(result);
      }
    }));
    return completer.future;
  }

  Future<FireLoginResult> login(String email, String password) async {
    Completer<FireLoginResult> completer = new Completer();
    this.js.login(email, password, allowInterop((jsResult)  {
      if(jsResult.user() != null) {
        var result = FireLoginResult();
        result.user = jsResult.user();
        completer.complete(result);
      } else {
        var result = FireLoginResult();
        result.error = DecideError(jsResult.error().code(), jsResult.error().message());
        completer.complete(result);
      }
    }));

    return completer.future;
  }
}

File: site/web/assets/js/decidefire.js
let getUser = function(uid, callback) {
    firebase.database().ref("users").child(uid).once("value", (snap) => {
        if(snap.val() != null) {
            let user = snap.val();
            user.uid = uid;
            if(user.groups === undefined) {
                user.groups = {};
            }
            callback(user);
        } else {
            callback({uid: uid, name: "Unknown Name", groups: {}});
        }
    });
};


function DecideFireJS() {
}

DecideFireJS.prototype.getKeysForObject = function(obj) {
    return Object.keys(obj);
};

DecideFireJS.prototype.push = function(path, obj, callback) {
    let ref = firebase.database().ref(path).push();
    ref.set(obj).then(() => {
        callback(ref.key);
    });
};

DecideFireJS.prototype.save = function(path, obj, callback) {
    firebase.database().ref(path).set(obj).then(() => {
        callback();
    });
};

DecideFireJS.prototype.get = function(path, callback) {
    console.log("getting at path " + path);
    firebase.database().ref(path).once("value", (snap) => {
        callback(snap.val());
    });
};

DecideFireJS.prototype.logout = function() {
    firebase.auth().signOut();
};

DecideFireJS.prototype.currentUser = function(callback) {
    let calledAlready = false;
    firebase.auth().onAuthStateChanged(function(user) {
        if(calledAlready) {
            return;
        } else {
            calledAlready = true;
        }
        if (user != null) {
            getUser(user.uid, callback);
        } else {
            callback(null);
        }
    });
};

DecideFireJS.prototype.checkEmail = function(email, callback) {
    console.log("checking " + email);
    firebase.auth().fetchSignInMethodsForEmail(email).then((methods) => callback(methods, "")).catch((error) => callback([], error));
};

DecideFireJS.prototype.signup = function(email, password, displayName, callback) {
    firebase.auth().setPersistence(firebase.auth.Auth.Persistence.LOCAL)
        .then(function() {
             return firebase.auth().createUserWithEmailAndPassword(email, password).then((cred) => {
                let uid = cred.user.uid;
                 firebase.auth().signInWithEmailAndPassword(email, password).then((cred) => {
                     callback(new DecideFireLoginResultJS({uid: uid, name: displayName, groups: {}}));
                     firebase.database().ref("users").child(uid).set({
                         name: displayName,
                         groups: {}
                     });
                 });
            });
        })
        .catch(function(error) {
            callback(new DecideFireLoginResultJS(null, new DecideFireErrorJS(error)));
        });

};

DecideFireJS.prototype.login = function(email, password, callback) {
    firebase.auth().setPersistence(firebase.auth.Auth.Persistence.LOCAL)
        .then(function() {
            return firebase.auth().signInWithEmailAndPassword(email, password).then((cred) => {
                let uid = cred.user.uid;
                getUser(uid, (user) => {
                    callback(new DecideFireLoginResultJS(user));
                });
            });
        })
        .catch(function(error) {
            callback(new DecideFireLoginResultJS(null, new DecideFireErrorJS(error)));
        });
};

function DecideFireUserJS(uid, name) {
    this._uid = uid;
    this._name = name;
}

DecideFireUserJS.prototype.uid = function() {
    return this._uid;
};

DecideFireUserJS.prototype.name = function() {
    return this._name;
};

function DecideFireErrorJS(error) {
    this.error = error;
}

DecideFireErrorJS.prototype.code = function() {
    return this.error.code;
};

DecideFireErrorJS.prototype.message = function() {
    return this.error.message;
};

function DecideFireLoginResultJS(user, error) {
    this._user = user;
    this._error = error;
}

DecideFireLoginResultJS.prototype.user = function() {
    return this._user;
};

DecideFireLoginResultJS.prototype.error = function() {
    return this._error;
};
